src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsRefactoringProvider.cs

```diff
@@ -34,25 +35,23 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
 
             // Check pasted text span for missing imports
             var addMissingImportsService = document.GetLanguageService<IAddMissingImportsFeatureService>();
-            var hasMissingImports = await addMissingImportsService.HasMissingImportsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
 
-            if (!hasMissingImports)
+            var analysis = await addMissingImportsService.AnalyzeAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (!analysis.CanAddMissingImports)
             {
                 return;
             }
 
             var addImportsCodeAction = new AddMissingImportsCodeAction(
                 CodeActionTitle,
-                cancellationToken => AddMissingImportsAsync(document, textSpan, cancellationToken));
-            context.RegisterRefactoring(addImportsCodeAction, textSpan);
-        }
+                async cancellationToken =>
+                {
+                    var modifiedDocument = await addMissingImportsService.AddMissingImportsAsync(analysis, cancellationToken).ConfigureAwait(false);
+                    RoslynDebug.AssertNotNull(modifiedDocument);
+                    return modifiedDocument.Project.Solution;
+                });
```

Comment:
i would undo this.  in general, for clarity, we always have the code-action body be a separate method.